File: greedy/programmers/practice_kit_1.py
def solution(n, lost, reserve):
    for i in reserve:
        if i in lost:
            reserve.remove(i)
            lost.remove(i)
    
    for i in reserve:
        if i in lost: lost.remove(i)
        if i - 1 in lost:
            lost.remove(i - 1)
        elif i + 1 in lost:
            lost.remove(i + 1)
    
    return n - len(lost)

def solution1(n, lost, reserve):
    set_lost = set(lost)-set(reserve)
    set_reserve = set(reserve)-set(lost)
            
    for i in set_reserve:
        if i - 1 in set_lost:
            set_lost.remove(i - 1)
        elif i + 1 in set_lost:
            set_lost.remove(i + 1)
    
    return n - len(set_lost)

def solution2(n, lost, reserve):
    remove = [x for x in lost if x in reserve]
    for x in remove:
        lost.remove(x)
        reserve.remove(x)

    for x in reserve:
        if x-1 in lost:
            lost.remove(x-1)
        elif x+1 in lost:
            lost.remove(x+1)
            
    return n-len(lost)

print(solution(3, [3], [1]))

File: greedy/programmers/practice_kit_1.java
package greedy.programmers;

class Solution {
    public int solution(int n, int[] lost, int[] reserve) {
        int[] people = new int[n];
        int answer = 0;

        for (int l : lost)
            people[l-1]--;
        for (int r : reserve)
            people[r-1]++;

        for (int i = 0; i < people.length; i++) {
            if (people[i] >= 1) {
                if (i-1 >= 0 && people[i-1] == -1) {
                    people[i]--;
                    people[i-1]++;
                } else if (i+1 < people.length && people[i+1] == -1) {
                    people[i]--;
                    people[i+1]++;
                }
            }
        }

        for (int i = 0; i < people.length; i++) {
            if (people[i] >= 0)
                answer++;
        }

        return answer;
    }
}

public class practice_kit_1 {
    public static void main(String[] args) {
        Solution s = new Solution();
        int n = 5;
        int[] lost = {2, 4};
        int[] reserve = {1, 3, 5};
        
        System.out.println(s.solution(n, lost, reserve));
    }

    
}